### ISSUE TITLE
Custom cpp configration include paths bug.

### ISSUE THREAD

**Issue by Polytoo** [microsoft/vscode-cpptools#3584]
Issue Type: <b>Bug</b>

I use my custom "Switch" configuration, but it searches macFrameworkPath for sdl.h file automatically, and it comes out 2 definitions of sdl.h file in 2 different dirs. 

One is in "/Library/Frameworks",  another in "${DEVKITPRO}/portlibs/switch/SDL2".

I just don't need to search the Mac default FrameworkPath.

here is my c_cpp_properties.json:
``` json
{
    "configurations": [
        {
            "name": "Mac",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [],
            "macFrameworkPath": [
                "/System/Library/Frameworks",
                "/Library/Frameworks"
            ],
            "compilerPath": "/usr/bin/clang",
            "cStandard": "c11",
            "cppStandard": "c++17",
            "intelliSenseMode": "clang-x64"
        },
        {
            "name": "Switch",
            "includePath": [            
                "${DEVKITPRO}/portlibs/switch/**",
                "${DEVKITPRO}/libnx/**",
                "${workspaceFolder}/**"
            ],
            "defines": [
                "__SWITCH__"
            ],
            "compilerPath": "${DEVKITPRO}/devkitA64/bin/aarch64-none-elf-g++",
            "cStandard": "c11",
            "cppStandard": "c++14",
            "intelliSenseMode": "gcc-x64"
        }
    ],
    "version": 4
}
```

Extension version: 0.23.0-insiders2
VS Code version: Code 1.33.1 (51b0b28134d51361cf996d2f0a1c698247aeabd8, 2019-04-11T08:22:55.268Z)
OS version: Darwin x64 17.7.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz (8 x 2000)|
|GPU Status|2d_canvas: unavailable_software<br>checker_imaging: disabled_off<br>flash_3d: unavailable_off<br>flash_stage3d: unavailable_off<br>flash_stage3d_baseline: unavailable_off<br>gpu_compositing: unavailable_off<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: enabled<br>rasterization: unavailable_off<br>surface_synchronization: enabled_on<br>video_decode: unavailable_off<br>webgl: unavailable_off<br>webgl2: unavailable_off|
|Load (avg)|2, 2, 2|
|Memory (System)|16.00GB (4.17GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

</details>

**Comment by bobbrow** [comment]
What happens when you set `"macFrameworkPath": []` in your configuration?  Or, if you run your compiler with `-v -E -x c++ -` as arguments, does it list the framework paths as part of the include search?

**Comment by kisielk** [comment]
I have the same problem with a different compiler. I have `compilerPath` set to `/usr/local/bin/arm-none-eabi-g++` and also set `"macFrameworkPath": []` but it looks like the intellisense engine is still querying the default g++ compiler for its paths and putting those first. 

I get the following in the debug log:

```
Folder: /usr/local/Cellar/gcc/6.3.0_1/lib/gcc/6/gcc/x86_64-apple-darwin16.3.0/6.3.0/include/ will be indexed
  Folder: /usr/local/include/ will be indexed
  Folder: /usr/local/Cellar/gcc/6.3.0_1/include/ will be indexed
  Folder: /usr/local/Cellar/gcc/6.3.0_1/lib/gcc/6/gcc/x86_64-apple-darwin16.3.0/6.3.0/include-fixed/ will be indexed
  Folder: /Library/Frameworks/Mono.framework/Versions/5.0.0/include/ will be indexed
  Folder: /Library/Frameworks/Panda.framework/Headers/ will be indexed
  Folder: /Library/Frameworks/Jackservermp.framework/Headers/ will be indexed
  Folder: /Library/Frameworks/Jacknet.framework/Headers/ will be indexed
  Folder: /Library/Frameworks/Jackmp.framework/Headers/ will be indexed
  Folder: /Library/Frameworks/DYMO/ will be indexed
  Folder: /Library/Frameworks/Resources/ will be indexed
  Folder: /Library/Frameworks/Versions/ will be indexed
  Folder: /Library/Frameworks/Headers/ will be indexed
 ...
```
where ... is the contents of "includePath".

Interestingly though, I have
```
                "/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/include",
                "/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/include-fixed",
                "/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/include",
```

as part of `"includePath"` but those paths are not listed there. I generated those using `arm-none-eabi-g++ -v -E -x c++ -`

**Comment by bobbrow** [comment]
@kisielk can you share your `c_cpp_properties.json` file?  The log output that shows `"Folder:"` may not match your `"includePath"` if the `"browse"` setting is being used.

**Comment by kisielk** [comment]
Here it is:
```{
    "configurations": [
        {
            "name": "STM32",
            "includePath": [
                "${workspaceFolder}",
                "${workspaceFolder}/Inc",
                "${workspaceFolder}/../common",
                "${workspaceFolder}/../../Drivers/STM32F4xx_HAL_Driver/Inc",
                "${workspaceFolder}/../../Drivers/CMSIS/CMSIS/Core/Include",
                "${workspaceFolder}/../../Drivers/STM32_USB_Device_Library/Core/Inc",
                "${workspaceFolder}/../../Drivers/STM32_USB_Device_Library/Class/CDC/Inc",
                "${workspaceFolder}/../../Drivers/CMSIS/Device/ST/STM32F4xx/Include/",
                "${workspaceFolder}/../../lib",
                "${workspaceFolder}/../../lib/intellijel",
                "${workspaceFolder}/../../lib/intellijel/rtos",
                "${workspaceFolder}/../../FatFS/src",
                "${workspaceFolder}/../../FreeRTOS/Source/include",
                "${workspaceFolder}/tests",
                "/usr/local/include",
                "/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/include",
                "/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/include-fixed",
                "/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/include",
                "${workspaceFolder}/../../FreeRTOS/Source/portable/CCS/ARM_CM4F"
            ],
            "defines": [
                "STM32F4xx",
                "STM32F405xx",
                "USE_FREERTOS"
            ],
            "macFrameworkPath": [],
            "intelliSenseMode": "gcc-x64",
            "browse": {
                "path": [
                    "${workspaceFolder}",
                    "${workspaceFolder}/../common",
                    "${workspaceFolder}/../../Drivers/STM32F4xx_HAL_Driver/Inc",
                    "${workspaceFolder}/../../Drivers/STM32F4xx_HAL_Driver/Src",
                    "${workspaceFolder}/../../Drivers/STM32_USB_Device_Library/Core/Inc",
                    "${workspaceFolder}/../../Drivers/STM32_USB_Device_Library/Core/Src",
                    "${workspaceFolder}/../../Drivers/STM32_USB_Device_Library/Class/CDC/Inc",
                    "${workspaceFolder}/../../Drivers/STM32_USB_Device_Library/Class/CDC/Src",
                    "${workspaceFolder}/../../lib",
                    "${workspaceFolder}/../../lib/intellijel",
                    "${workspaceFolder}/../../lib/intellijel/rtos",
                    "${workspaceFolder}/../../FatFS/src",
                    "${workspaceFolder}/../../FreeRTOS/Source",
                    "${workspaceFolder}/tests",
                    "/usr/local/include",
                    "/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/include",
                    "/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/include-fixed",
                    "/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/include"
                ],
                "limitSymbolsToIncludedHeaders": true,
                "databaseFilename": ""
            },
            "compilerPath": "/usr/local/bin/arm-none-eabi-g++",
            "cStandard": "c11",
            "cppStandard": "c++17",
            "configurationProvider": "vector-of-bool.cmake-tools",
            "compileCommands": "${workspaceFolder}/build/compile_commands.json"
        }
    ],
    "version": 4
}```

**Comment by bobbrow** [comment]
The paths are coming from `"configurationProvider": "vector-of-bool.cmake-tools"`.  When you use a `configurationProvider` our extension delegates configuration to the other extension.  We only use the `includePath` and `browse` settings when CMake Tools fails to provide the configuration.

**Comment by kisielk** [comment]
Hm I see, I've got the cmake extension configured to use the same compiler so it's still weird that it would getting system paths in there...

**Comment by bobbrow** [comment]
Can you search the debug log for "Custom browse configuration received" and see what CMake Tools is sending for the `compilerPath`?

**Comment by bobbrow** [comment]
It could be that the wrong Kit is set?

An example of Kits on my machine:
![image](https://user-images.githubusercontent.com/12818240/66852811-a8265b00-ef32-11e9-9c71-7014bb97edda.png)

**Comment by kisielk** [comment]
It seems to have the right `compilerPath`:
```
Custom configurations received:
  uri: file:///Users/kamil/intellijel/eurorack/tetrapad/expander/Src/tete.cpp
  config: {
  "defines": [
    "ARM_MATH_CM4",
    "STM32F405RGTx",
    "STM32F405xx",
    "STM32F4xx",
    "USE_FREERTOS",
    "USE_FULL_LL_DRIVER",
    "DEBUG"
  ],
  "standard": "c++17",
  "includePath": [
    "/Users/kamil/intellijel/eurorack/tetrapad/expander/Inc",
    "/Users/kamil/intellijel/eurorack/tetrapad/expander",
    "/Users/kamil/intellijel/eurorack/tetrapad/expander/../common",
    "/Users/kamil/intellijel/eurorack/cmake/../FatFs/src",
    "/Users/kamil/intellijel/eurorack/cmake/../FreeRTOS/Source/include",
    "/Users/kamil/intellijel/eurorack/cmake/../FreeRTOS/Source/portable/GCC/ARM_CM4F",
    "/Users/kamil/intellijel/eurorack/cmake/../Drivers/STM32F4xx_HAL_Driver/Inc",
    "/Users/kamil/intellijel/eurorack/cmake/../Drivers/CMSIS/Device/ST/STM32F4xx/Include",
    "/Users/kamil/intellijel/eurorack/cmake/../Drivers/CMSIS/CMSIS/Include",
    "/Users/kamil/intellijel/eurorack/lib/intellijel/.."
  ],
  "intelliSenseMode": "clang-x64",
  "compilerPath": "arm-none-eabi-g++"
}
```

I have the kit set to `GCC for arm-none-eabi 8.2.1`

**Comment by sean-mcmanus** [comment]
Looks like the intelliSenseMode got changed to "clang-x64" incorrectly? It's possible there was some failure querying the compiler. Can you see what defines are shown when you enable logging? https://code.visualstudio.com/docs/cpp/enable-logging-cpp

**Comment by kisielk** [comment]
I think the problem is the CMake tools extension is not providing the full compiler path, and the C++ extension doesn't search for it in `/usr/local/bin` I see this:
```
Queueing IntelliSense update for files in translation unit of: /Users/kamil/intellijel/eurorack/tetrapad/expander/Src/tete.cpp
Error squiggle count: 0
cpptools/didChangeCustomConfiguration
Compiler in "compilerPath" property not found: arm-none-eabi-g++
Attempting to get defaults from compiler found on the machine: '/usr/local/bin/gcc-6
```

**Comment by kisielk** [comment]
Ok, I think I have got it working now...
I had to define my own kit for the CMake Tools extension. The default one it scanned looked like this:

```
  {
    "name": "GCC for arm-none-eabi 8.2.1",
    "compilers": {
      "C": "/usr/local/bin/arm-none-eabi-gcc-8.2.1"
    }
  },
```

Note that there's no `"CXX"` key (other kits have this). I created my own kit that looks like:

```
  {
    "name": "GCC for arm-none-eabi",
    "compilers": {
      "C": "/usr/local/bin/arm-none-eabi-gcc",
      "CXX": "/usr/local/bin/arm-none-eabi-g++"
    }
  }
```

and now the IntelliSense engine seems to be working correctly.

It looks like this may be happening because there's a `/usr/local/bin/arm-none-eabi-gcc-8.2.1` binary, but no corresponding `/usr/local/bin/arm-none-eabi-g++-8.2.1` binary.
